pkg/cmd/certregenerationcontroller/csrapprover.go

```diff
@@ -0,0 +1,298 @@
+package certregenerationcontroller
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	certificatesv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/csr"
+	"github.com/openshift/library-go/pkg/operator/events"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const (
+	controllerName       = "master-nodes-csr-approver-controller"
+	bootstrapperUsername = "system:serviceaccount:openshift-machine-config-operator:node-bootstrapper"
+)
+
+var (
+	mcoUserGroups = sets.NewString(
+		"system:authenticated",
+		"system:serviceaccounts",
+		"system:serviceaccounts:openshift-machine-config-operator",
+	)
+	kubeletUserGroups = sets.NewString("system:authenticated", "system:nodes")
+)
+
+type ControlPlaneCSRController struct {
+	controller   factory.Controller
+	nodeApprover *ControlPlaneNodeApprover
+}
+
+func NewControlPlaneCSRController(
+	ctx context.Context,
+	kubeClient *kubernetes.Clientset,
+	kubeInformers informers.SharedInformerFactory,
+	eventRecorder events.Recorder,
+) (*ControlPlaneCSRController, error) {
+	klog.Infof("Starting ControlPlaneCSRController")
+	controlPlaneNodeApprover, err := NewControlPlaneNodeApprover(ctx, kubeClient, kubeInformers, eventRecorder)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new control plane node approver: %w", err)
+	}
+	return &ControlPlaneCSRController{
+		controller: csr.NewCSRApproverController(
+			controllerName,
+			nil,
+			kubeClient.CertificatesV1().CertificateSigningRequests(),
+			kubeInformers.Certificates().V1().CertificateSigningRequests(),
+			csr.NewLabelFilter(labels.NewSelector()),
```

Comment:
if you only want to approve the first kubelet client certificate, this needs to filter on
1. the username of the CSR requestor must be the bootstrapping service account
2. the CSR must be for the right signer
3. the CSR must be for a kubelet shaped client certificates